pkg/scheduler/api/hyper_node_info.go

```diff
@@ -0,0 +1,463 @@
+}
+
+// RealNodesSet returns a deep copy of the map that stores the sets of real nodes for each HyperNode.
+// This ensures that the returned map is independent of the original, preventing unintended modifications.
+func (hni *HyperNodesInfo) RealNodesSet() map[string]sets.Set[string] {
+	copiedRealNodesSet := make(map[string]sets.Set[string], len(hni.realNodesSet))
+	for name, value := range hni.realNodesSet {
+		copiedRealNodesSet[name] = value.Clone() // Clone the set to ensure a deep copy.
+	}
+
+	return copiedRealNodesSet
+}
+
+// ParentMap returns the map of parent-child relationships between HyperNodes.
+func (hni *HyperNodesInfo) ParentMap() map[string]string {
+	return hni.parentMap
+}
+
+// DeleteHyperNode deletes a HyperNode from the cache and update hyperNode tree.
+func (hni *HyperNodesInfo) DeleteHyperNode(name string) error {
+	hni.markHyperNodeIsDeleting(name)
+	if err := hni.updateAncestors(name); err != nil {
+		return err
+	}
+	hni.removeParent(name)
+
+	// We can safely delete hyperNode after updated ancestors.
+	hni.deleteHyperNode(name)
+	return nil
+}
+
+func (hni *HyperNodesInfo) UpdateHyperNode(hn *topologyv1alpha1.HyperNode) error {
+	hni.updateHyperNode(hn)
+	return hni.updateAncestors(hn.Name)
+}
+
+func (hni *HyperNodesInfo) BuildHyperNodeCache(hn *HyperNodeInfo, processed sets.Set[string], ancestors sets.Set[string], nodes []*corev1.Node) error {
+	// Check for cyclic dependency.
+	if ancestors.Has(hn.Name) {
+		hni.setReady(false)
+		return fmt.Errorf("cyclic dependency detected: HyperNode %s is already in the ancestor chain %v", hn.Name, sets.List(ancestors))
+	}
+	// Pruning operation: skip if hyperNode already processed.
+	if processed.Has(hn.Name) {
+		return nil
+	}
+
+	ancestors.Insert(hn.Name)
+	defer ancestors.Delete(hn.Name)
+
+	if hni.HyperNodeIsDeleting(hn.Name) {
+		klog.InfoS("HyperNode is being deleted", "name", hn.Name)
+		return nil
+	}
+
+	for _, member := range hn.HyperNode.Spec.Members {
+		switch member.Type {
+		case topologyv1alpha1.MemberTypeNode:
+			if _, ok := hni.realNodesSet[hn.Name]; !ok {
+				hni.realNodesSet[hn.Name] = sets.New[string]()
+			}
+			members := hni.getMembers(member.Selector, nodes)
+			klog.V(5).InfoS("Get members of hyperNode", "name", hn.Name, "members", members)
+			hni.realNodesSet[hn.Name] = hni.realNodesSet[hn.Name].Union(members)
+
+		case topologyv1alpha1.MemberTypeHyperNode:
+			// HyperNode member does not support regex match.
+			memberName := hni.exactMatchMember(member.Selector)
+			if memberName == "" {
+				continue
+			}
+
+			if err := hni.setParent(hn, memberName); err != nil {
+				return err
+			}
+
+			memberHn, ok := hni.hyperNodes[memberName]
+			if !ok {
+				klog.InfoS("HyperNode not exists in cache, maybe not created or not be watched", "name", memberName, "parent", hn.Name)
+				continue
+			}
+
+			// Recursively build cache for the member HyperNode.
+			if err := hni.BuildHyperNodeCache(memberHn, processed, ancestors, nodes); err != nil {
```

Comment:
Only the realNodesSet on the ancestors chain will be reset. Can the HyperNode that is not on the ancestor chain don't need to recursively enter BuildHyperNodeCache? Just check whether the realNodesSet of this member is not empty.

Reply:
Otherwise there will be some repetitive builds, like: 
![image](https://github.com/user-attachments/assets/654f72f7-4e25-45f1-bae2-99cdb5b2d2ac)

Reply:
done